### ISSUE TITLE
gradients with general shape

### ISSUE THREAD

**Issue by YihanWang617** [pytorch/pytorch#48417]
## 🚀 Feature
back-propagate the gradients with general shapes.

## Motivation

I am trying to use the backward engine implemented in pytorch to implement a kind of "general gradients", which also propagates the tensors backward on the graph but the 'general gradients' can support more general shapes. For example, for a linear operation which accepts x \in R^d and outputs y \in R^m, the gradient of y w.r.t. x will be of the shape of [d, m].

However, the implementation of pytorch strictly constraints the shape of gradients. I have personally modified the source code of pytorch to meet my own use, but I am wondering if pytorch can officially support this feature. I think the backward process can be also useful besides calculating derivatives.

## Pitch
Here is a small example.
```
import torch
x = torch.tensor([1, 2])
y = torch.tensor([2, 3])

z = x + y
```
To calculate the Jacobin matrix of z, I can override the backward function of torch.add. However, when I call `z.backward([[1, 1], [1, 1]])`, an error will occur due to the shape constraints in pytorch.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

**Comment by ngimel** [comment]
Can you tell a bit more about your usecase and why would you need the gradients of the different shape than the corresponding tensor?

**Comment by YihanWang617** [comment]
> Can you tell a bit more about your usecase and why would you need the gradients of the different shape than the corresponding tensor?

Hi ngimel,

For example, I want to compute the jocabian matrix in a single backward, which is possible by backward propagating a matrix of size [output, input]. But this shape will not be acceptable in pytorch.

**Comment by zou3519** [comment]
@YihanWang617 You can use torch.autograd.functional.jacobian to compute the full jacobian: https://pytorch.org/docs/stable/autograd.html.

**Comment by albanD** [comment]
Hi,

The problem is that the formulas we define for the backward pass only handle gradients, not batch of gradients (which is effectively what you're trying to do).
So even though modifying the engine to allow this would be simple, it would also require updating all the formulas.

**Comment by soulitzer** [comment]
Closing because the batched grad use case is possible now: 
- On the main branch computing batched grads is now supported. Setting the `is_grads_batched` parameter to `True` will allow one extra batch dimension for your `grad_out`. 
- For more general use cases involving batching one can use `vmap(grad` in functorch https://github.com/pytorch/functorch.

Feel free to reopen if there is another use case of gradients with general shapes that is not covered
